File: app/src/main/java/jp/kentan/student_portal_plus/notification/NotificationController.java
package jp.kentan.student_portal_plus.notification;


import android.app.Notification;
import android.app.PendingIntent;
import android.content.Context;
import android.content.Intent;
import android.content.SharedPreferences;
import android.graphics.Bitmap;
import android.graphics.BitmapFactory;
import android.os.Build;
import android.support.v4.app.NotificationCompat;
import android.support.v4.app.NotificationManagerCompat;
import android.support.v4.content.res.ResourcesCompat;
import android.text.Spanned;
import android.util.Log;

import java.util.List;

import jp.kentan.student_portal_plus.ui.HomeActivity;
import jp.kentan.student_portal_plus.ui.LectureCancellationActivity;
import jp.kentan.student_portal_plus.ui.NewsActivity;
import jp.kentan.student_portal_plus.ui.LectureInformationActivity;
import jp.kentan.student_portal_plus.R;

public class NotificationController {

    private final static long[] VIBRATE = {0, 300, 300, 300};

    private final static String GROUP_KEY = "student_portal_plus";

    private final static int INBOX_LINE_LIMIT = 4;
    final static int ERROR_ID = -1;

    private static Bitmap LARGE_ICON;
    private static int APP_ICON;
    private static int[] SMALL_ICON = new int[3];

    private static int COLOR_ACCENT;
    private final boolean isVibrate, isLed;

    private NotificationManagerCompat mNotificationManager;
    private Context mContext;
    private SharedPreferences mPreference;

    private final static boolean IS_SUPPORT_NOTIFICATION_SUMMARY = (Build.VERSION.SDK_INT >= Build.VERSION_CODES.N);
    private final static boolean IS_SUPPORT_VECTOR_IMG           = (Build.VERSION.SDK_INT > Build.VERSION_CODES.KITKAT);
    private boolean isFirstNotify = true, isSummaryAvailable = false;

    private final static int SUMMARY_NOTIFY_ID = 0;
    private int mLastNotifyId = 1; //IS_SUPPORT_NOTIFICATION_SUMMARY
    private final int[] NOTIFY_ID = {1/* 授業関連連絡 */, 2/* 休講情報 */, 3/* 最新情報 */};
    private final int[] VIEWMODE  = {3/* 授業関連連絡 */, 4/* 休講情報 */, 5/* 最新情報 */};


    public NotificationController(Context context) {
        mContext = context.getApplicationContext();

        mPreference = mContext.getSharedPreferences("notification", Context.MODE_PRIVATE);

        isVibrate = mPreference.getBoolean("vibrate", true);
        isLed     = mPreference.getBoolean("led"    , true);

        mLastNotifyId = mPreference.getInt("last_notify_id", 1);

        if(mLastNotifyId <= SUMMARY_NOTIFY_ID) mLastNotifyId = 1;

        LARGE_ICON = BitmapFactory.decodeResource(mContext.getResources(), R.mipmap.ic_notify);

        //Check vector image support
        if (IS_SUPPORT_VECTOR_IMG) {
            APP_ICON = R.drawable.ic_menu_dashboard;

            SMALL_ICON[0] = R.drawable.ic_menu_class_schedule_change;
            SMALL_ICON[1] = R.drawable.ic_menu_class_cancel;
            SMALL_ICON[2] = R.drawable.ic_menu_news_events;
        } else {
            APP_ICON = R.mipmap.ic_app;

            SMALL_ICON[0] = R.mipmap.ic_class_schedule_change;
            SMALL_ICON[1] = R.mipmap.ic_class_cancel;
            SMALL_ICON[2] = R.mipmap.ic_news_events;
        }

        COLOR_ACCENT = ResourcesCompat.getColor(mContext.getResources(), R.color.colorAccent, null);

        mNotificationManager = NotificationManagerCompat.from(mContext);
    }

    public void cancel(int notifyId) {
        mNotificationManager.cancel(notifyId);
    }

    public void cancelAll() {
        mNotificationManager.cancelAll();

        SharedPreferences.Editor editor = mPreference.edit();
        editor.putInt("last_notify_id", 1);
        editor.apply();
    }

    void destroy() {
        SharedPreferences.Editor editor = mPreference.edit();
        editor.putInt("last_notify_id", mLastNotifyId);
        editor.apply();
    }

    void notify(Content.TYPE type, List<Content> list){
        if(list.size() <= 0) return;

        final int intType = type.ordinal();

        if (IS_SUPPORT_NOTIFICATION_SUMMARY) {
            createNotificationSummaryIfNeed();

            final String subText = Content.NAME[intType];

            for(Content content : list){
                NotificationCompat.Builder builder = new NotificationCompat.Builder(mContext)
                        .setVisibility(NotificationCompat.VISIBILITY_PUBLIC)
                        .setPriority(NotificationCompat.PRIORITY_HIGH)
                        .setSmallIcon(SMALL_ICON[intType])
                        .setGroup(GROUP_KEY)
                        .setSubText(subText)
                        .setContentTitle(content.getTitle())
                        .setContentText(content.getText())
                        .setContentIntent(createPendingIntent(content));

                mNotificationManager.notify(mLastNotifyId++, build(builder));
            }
        } else {
            Intent intent = new Intent(mContext, HomeActivity.class);
            intent.putExtra("view_mode", VIEWMODE[intType]);
            PendingIntent contentIntent = PendingIntent.getActivity(mContext, NOTIFY_ID[intType], intent, 0);

            NotificationCompat.Builder builder = getNotificationBuilder(type, Content.getInboxStyleTitle(type, list.size()), list.get(0).getInboxStyleText(), contentIntent);

            NotificationCompat.InboxStyle inbox = getInboxStyle(Content.NAME[intType], list);
            builder.setStyle(inbox);

            mNotificationManager.notify(NOTIFY_ID[intType], build(builder));
        }
    }

    void notifyAutoFetchFailed(String errorMessage) {
        mLastNotifyId++;

        PendingIntent contentIntent = PendingIntent.getActivity(mContext, mLastNotifyId, new Intent(mContext, HomeActivity.class), 0);

        Intent service = new Intent(mContext, NotificationService.class);
        service.putExtra("retry", true);

        PendingIntent serviceIntent = PendingIntent.getService(mContext, ERROR_ID, service, PendingIntent.FLAG_UPDATE_CURRENT);
        NotificationCompat.Action actionRetry = new NotificationCompat.Action.Builder(R.drawable.ic_refresh, "再試行", serviceIntent).build();

        NotificationCompat.Builder builder = new NotificationCompat.Builder(mContext)
                .setVisibility(NotificationCompat.VISIBILITY_PUBLIC)
                .setCategory(NotificationCompat.CATEGORY_ERROR)
                .setPriority(NotificationCompat.PRIORITY_LOW)
                .setSmallIcon(APP_ICON)
                .setContentTitle("自動受信失敗")
                .setContentText(errorMessage)
                .setContentIntent(contentIntent);

        if(IS_SUPPORT_VECTOR_IMG){
            builder.addAction(actionRetry);
        }

        Notification notification = builder.build();
        notification.flags = Notification.FLAG_AUTO_CANCEL;

        mNotificationManager.notify(ERROR_ID, notification);
    }

    private void createNotificationSummaryIfNeed() {
        if (!IS_SUPPORT_NOTIFICATION_SUMMARY || isSummaryAvailable) return;
        isSummaryAvailable = true;

        final PendingIntent contentIntent = PendingIntent.getActivity(mContext, SUMMARY_NOTIFY_ID, new Intent(mContext, HomeActivity.class), 0);

        Notification summaryNotification = new NotificationCompat.Builder(mContext)
                .setVisibility(NotificationCompat.VISIBILITY_PUBLIC)
                .setSmallIcon(APP_ICON)
                .setColor(COLOR_ACCENT)
                .setGroup(GROUP_KEY)
                .setGroupSummary(true)
                .setAutoCancel(true)
                .setContentIntent(contentIntent)
                .build();
        mNotificationManager.notify(SUMMARY_NOTIFY_ID, summaryNotification);
    }

    private PendingIntent createPendingIntent(Content content){
        final Intent intent;

        switch (content.getType()){
            case LECTURE_INFO:
                intent = new Intent(mContext, LectureInformationActivity.class);
                break;
            case LECTURE_CANCEL:
                intent = new Intent(mContext, LectureCancellationActivity.class);
                break;
            case NEWS:
                intent = new Intent(mContext, NewsActivity.class);
                break;
            default:
                return null;
        }

        intent.setFlags(Intent.FLAG_ACTIVITY_CLEAR_TOP | Intent.FLAG_ACTIVITY_NEW_TASK);
        intent.putExtra("hash", content.getHash());
        intent.putExtra("notify_id", mLastNotifyId);
        intent.putExtra("notification", true);

        Log.d("Notify", mLastNotifyId + ":" + content.getTitle());

        return PendingIntent.getActivity(mContext, mLastNotifyId, intent, PendingIntent.FLAG_UPDATE_CURRENT);
    }

    private NotificationCompat.InboxStyle getInboxStyle(String title, List<Content> list) {
        final int size = list.size();

        NotificationCompat.InboxStyle inboxStyle = new NotificationCompat.InboxStyle();
        inboxStyle.setBigContentTitle(title);

        for (int i = 0; i < size; ++i) {
            inboxStyle.addLine(list.get(i).getInboxStyleText());
            if (i >= INBOX_LINE_LIMIT - 1) break;
        }

        final int more = size - INBOX_LINE_LIMIT;
        if (more > 0) {
            inboxStyle.setSummaryText("他" + more + "件");
        }

        return inboxStyle;
    }

    private NotificationCompat.Builder getNotificationBuilder(Content.TYPE type, String title, Spanned text, PendingIntent intent) {
        return new NotificationCompat.Builder(mContext)
                .setVisibility(NotificationCompat.VISIBILITY_PUBLIC)
                .setPriority(NotificationCompat.PRIORITY_HIGH)
                .setLargeIcon(LARGE_ICON)
                .setSmallIcon(SMALL_ICON[type.ordinal()])
                .setColor(COLOR_ACCENT)
                .setContentTitle(title)
                .setContentText(text)
                .setContentIntent(intent);
    }

    private Notification build(NotificationCompat.Builder builder) {
        Notification notification;

        if (isFirstNotify) {
            isFirstNotify = false;

            if (isLed) builder.setLights(COLOR_ACCENT, 1000, 2000);
            builder.setVibrate((isVibrate) ? VIBRATE : new long[]{0});

            notification = builder.build();
            if (isLed) notification.flags = Notification.FLAG_SHOW_LIGHTS;
        } else {
            notification = builder.build();
        }

        notification.flags |= Notification.FLAG_AUTO_CANCEL;

        return notification;
    }
}

File: app/src/main/java/jp/kentan/studentportalplus/notification/NotificationController.kt
package jp.kentan.studentportalplus.notification

import android.app.Notification
import android.app.NotificationChannel
import android.app.NotificationManager
import android.app.PendingIntent
import android.content.Context
import android.content.ContextWrapper
import android.content.Intent
import android.graphics.BitmapFactory
import android.graphics.Typeface
import android.os.Build
import android.text.Spannable
import android.text.style.StyleSpan
import androidx.core.app.NotificationCompat
import androidx.core.app.NotificationManagerCompat
import androidx.core.content.ContextCompat
import androidx.core.text.set
import androidx.core.text.toSpannable
import jp.kentan.studentportalplus.R
import jp.kentan.studentportalplus.data.component.PortalContent
import jp.kentan.studentportalplus.data.component.PortalData
import jp.kentan.studentportalplus.ui.lecturecancel.detail.LectureCancelDetailActivity
import jp.kentan.studentportalplus.ui.lectureinfo.detail.LectureInfoDetailActivity
import jp.kentan.studentportalplus.ui.login.LoginActivity
import jp.kentan.studentportalplus.ui.main.FragmentType
import jp.kentan.studentportalplus.ui.main.MainActivity
import jp.kentan.studentportalplus.ui.notice.detail.NoticeDetailActivity
import jp.kentan.studentportalplus.util.getNotificationId
import jp.kentan.studentportalplus.util.isEnabledNotificationLed
import jp.kentan.studentportalplus.util.isEnabledNotificationVibration
import jp.kentan.studentportalplus.util.setNotificationId
import org.jetbrains.anko.clearTask
import org.jetbrains.anko.defaultSharedPreferences
import org.jetbrains.anko.newTask

class NotificationController(context: Context) : ContextWrapper(context) {

    companion object {
        const val NEWLY_CHANNEL_ID = "0_newly_channel" //新着通知
        private const val APP_CHANNEL_ID = "99_app_channel"

        private const val GROUP_KEY = "student_portal_plus"
        private const val SUMMARY_NOTIFICATION_ID = 0
        private const val ERROR_NOTIFICATION_ID = -1
        private const val INBOX_LINE_LIMIT = 4

        private val CAN_USE_VECTOR_DRAWABLE = Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP
        private val CAN_USE_SUMMARY = Build.VERSION.SDK_INT >= Build.VERSION_CODES.N

        private val VIBRATION_PATTERN = longArrayOf(0, 300, 300, 300)

        private val SMALL_APP_ICON = if (CAN_USE_VECTOR_DRAWABLE) R.drawable.ic_menu_dashboard else R.mipmap.ic_notification_app
        private val SMALL_ICON_MAP = if (CAN_USE_VECTOR_DRAWABLE) {
            mapOf(
                    PortalData.LECTURE_INFO to R.drawable.ic_menu_lecture_info,
                    PortalData.LECTURE_CANCEL to R.drawable.ic_menu_lecture_cancel,
                    PortalData.NOTICE to R.drawable.ic_menu_notice
            )
        } else {
            mapOf(
                    PortalData.LECTURE_INFO to R.mipmap.ic_notification_lecture_info,
                    PortalData.LECTURE_CANCEL to R.mipmap.ic_notification_lecture_cancel,
                    PortalData.NOTICE to R.mipmap.ic_notification_notice
            )
        }


        fun setupChannel(context: Context) {
            if (Build.VERSION.SDK_INT < Build.VERSION_CODES.O) {
                return
            }

            val color = ContextCompat.getColor(context, R.color.colorAccent)

            val newlyChannel = NotificationChannel(NEWLY_CHANNEL_ID, context.getString(R.string.name_newly_channel), NotificationManager.IMPORTANCE_DEFAULT).apply {
                enableLights(true)
                lightColor = color
                vibrationPattern = VIBRATION_PATTERN
                lockscreenVisibility = Notification.VISIBILITY_PUBLIC
            }

            val appChannel = NotificationChannel(APP_CHANNEL_ID, context.getString(R.string.app_name), NotificationManager.IMPORTANCE_LOW).apply {
                enableLights(true)
                lightColor = color
                lockscreenVisibility = Notification.VISIBILITY_PUBLIC
            }

            val manager = context.getSystemService(Context.NOTIFICATION_SERVICE) as NotificationManager
            manager.createNotificationChannels(listOf(newlyChannel, appChannel))
        }
    }

    private val notificationManager = NotificationManagerCompat.from(context)

    private val isEnabledVibration = context.defaultSharedPreferences.isEnabledNotificationVibration()
    private val isEnabledLed = context.defaultSharedPreferences.isEnabledNotificationLed()

    private val accentColor = ContextCompat.getColor(context, R.color.colorAccent)
    private val largeIcon by lazy(LazyThreadSafetyMode.NONE) { BitmapFactory.decodeResource(context.resources, R.mipmap.ic_notification_large) }

    private var isFirstNotify = true


    init {
        setupChannel(context)
    }

    fun notify(type: PortalData, contentList: List<PortalContent>) {
        if (contentList.isEmpty()) {
            return
        }

        val smallIcon = SMALL_ICON_MAP[type] ?: SMALL_APP_ICON
        val name = getString(type.nameResId)
        var id = defaultSharedPreferences.getNotificationId()

        if (CAN_USE_SUMMARY) {
            createSummaryNotification()

            val builder = NotificationCompat.Builder(applicationContext, NEWLY_CHANNEL_ID)

            contentList.forEach {
                builder.setVisibility(NotificationCompat.VISIBILITY_PUBLIC)
                        .setSmallIcon(smallIcon)
                        .setGroup(GROUP_KEY)
                        .setGroupAlertBehavior(NotificationCompat.GROUP_ALERT_SUMMARY)
                        .setGroupSummary(false)
                        .setSubText(name)
                        .setContentTitle(it.title)
                        .setContentText(it.text)
                        .setContentIntent(it.createIntent(type, id))
                        .setAutoCancel(true)

                notificationManager.notify(id, builder.build())

                if (++id >= Int.MAX_VALUE) {
                    id = 1
                }
            }
        } else {
            val title = if (contentList.size > 1) getString(R.string.title_content_inbox, contentList.size, name) else name

            val builder = NotificationCompat.Builder(applicationContext, NEWLY_CHANNEL_ID)
                    .setVisibility(NotificationCompat.VISIBILITY_PUBLIC)
                    .setPriority(NotificationCompat.PRIORITY_HIGH)
                    .setLargeIcon(largeIcon)
                    .setSmallIcon(smallIcon)
                    .setColor(accentColor)
                    .setContentTitle(title)
                    .setContentText(contentList.first().toInboxStyleText())
                    .setContentIntent(type.createIntent(id))
                    .setStyle(contentList.createInboxStyle(type))
                    .setAutoCancel(true)

            if (isFirstNotify) {
                builder.setVibrate(if (isEnabledVibration) VIBRATION_PATTERN else longArrayOf(0))

                if (isEnabledLed) {
                    builder.setLights(accentColor, 1000, 2000)
                }
            }

            notificationManager.notify(id, builder.build())

            if (++id >= Int.MAX_VALUE) {
                id = 1
            }
        }

        isFirstNotify = false
        defaultSharedPreferences.setNotificationId(id)
    }

    fun notifyError(message: String, isRequireLogin: Boolean = false) {
        val intent = if (isRequireLogin) {
            LoginActivity.createIntent(this, true)
        } else {
            MainActivity.createIntent(this)
        }

        val pendingIntent = PendingIntent.getActivity(this, ERROR_NOTIFICATION_ID, intent, PendingIntent.FLAG_UPDATE_CURRENT)

        val builder = NotificationCompat.Builder(applicationContext, APP_CHANNEL_ID)
                .setVisibility(NotificationCompat.VISIBILITY_PUBLIC)
                .setCategory(NotificationCompat.CATEGORY_ERROR)
                .setPriority(NotificationCompat.PRIORITY_HIGH)
                .setColor(ContextCompat.getColor(this, R.color.red_600))
                .setSmallIcon(SMALL_APP_ICON)
                .setSubText(getString(R.string.sub_text_sync_error))
                .setContentTitle(getString(if (isRequireLogin) R.string.title_require_login else R.string.title_error))
                .setContentText(message)
                .setContentIntent(pendingIntent)
                .setAutoCancel(true)

        if (CAN_USE_VECTOR_DRAWABLE) {
            val retryService = PendingIntent.getService(this,
                    ERROR_NOTIFICATION_ID,
                    Intent(this, RetryActionService::class.java),
                    PendingIntent.FLAG_UPDATE_CURRENT)
            val retryAction = NotificationCompat.Action.Builder(R.drawable.ic_retry, getString(R.string.name_retry), retryService).build()

            builder.addAction(retryAction)
        }

        notificationManager.notify(ERROR_NOTIFICATION_ID, builder.build())
    }

    fun cancelErrorNotification() {
        notificationManager.cancel(ERROR_NOTIFICATION_ID)
    }

    private fun createSummaryNotification() {
        if (!CAN_USE_SUMMARY || !isFirstNotify) {
            return
        }

        val builder = NotificationCompat.Builder(applicationContext, NEWLY_CHANNEL_ID)
                .setVisibility(NotificationCompat.VISIBILITY_PUBLIC)
                .setSmallIcon(SMALL_APP_ICON)
                .setColor(accentColor)
                .setGroup(GROUP_KEY)
                .setGroupSummary(true)
                .setVibrate(if (isEnabledVibration) VIBRATION_PATTERN else longArrayOf(0))
                .setOnlyAlertOnce(true)
                .setAutoCancel(true)

        if (isEnabledLed) {
            builder.setLights(accentColor, 1000, 2000)
        }

        notificationManager.notify(SUMMARY_NOTIFICATION_ID, builder.build())
    }

    private fun PortalContent.createIntent(type: PortalData, code: Int): PendingIntent? {
        val context = this@NotificationController

        val intent = when (type) {
            PortalData.LECTURE_INFO -> LectureInfoDetailActivity.createIntent(context, id)
            PortalData.LECTURE_CANCEL -> LectureCancelDetailActivity.createIntent(context, id)
            PortalData.NOTICE -> NoticeDetailActivity.createIntent(context, id)
            else -> return null
        }

        return PendingIntent.getActivity(context, code, intent.clearTask().newTask(), PendingIntent.FLAG_UPDATE_CURRENT)
    }

    private fun PortalData.createIntent(id: Int): PendingIntent {
        val context = this@NotificationController

        val fragment = when (this) {
            PortalData.NOTICE -> FragmentType.NOTICE
            PortalData.LECTURE_INFO -> FragmentType.LECTURE_INFO
            PortalData.LECTURE_CANCEL -> FragmentType.LECTURE_CANCEL
            PortalData.MY_CLASS -> FragmentType.DASHBOARD // not support
        }

        val intent = MainActivity.createIntent(context, fragment = fragment)
        return PendingIntent.getActivity(context, id, intent, PendingIntent.FLAG_UPDATE_CURRENT)
    }

    private fun List<PortalContent>.createInboxStyle(type: PortalData): NotificationCompat.InboxStyle {
        val inboxStyle = NotificationCompat.InboxStyle()
                .setBigContentTitle(getString(type.nameResId))

        take(INBOX_LINE_LIMIT).forEach {
            inboxStyle.addLine(it.toInboxStyleText())
        }

        val moreContentSize = size - INBOX_LINE_LIMIT
        if (moreContentSize > 0) {
            inboxStyle.setSummaryText(getString(R.string.summary_text_inbox, moreContentSize))
        }

        return inboxStyle
    }

    private fun PortalContent.toInboxStyleText(): Spannable {
        val text = "$title $text".toSpannable()
        text[0..title.length] = StyleSpan(Typeface.BOLD)
        return text
    }
}